[RAM] Adds Bulk Edit API to rulesClient

vitaliidm opened the pull request:
Addresses
- https://github.com/elastic/kibana/issues/124715

## Summary

- adds bulkEdit method to rulesClient
- adds multi_terms bucket aggregations to savedObjectClient
- adds createPointInTimeFinderAsInternalUser to encryptedSavedObjectClient
- adds alerting API for bulkEdit
```bash
curl --location --request POST 'http://localhost:5601/kbn/internal/alerting/rules/_bulk_edit' \
--header 'kbn-xsrf: reporting' \
--header 'Authorization: Basic ZWxhc3RpYzpjaGFuZ2VtZQ==' \
--header 'Content-Type: application/json' \
--data-raw '{
    "ids": ["4cb80374-b5c7-11ec-8f1e-adaa7d7d57e5"],
    "operations":  [{
        "operation": "add",
        "field": "tags",
        "value": ["foo"]
    }]
}'
```
### Checklist

Delete any items that are not applicable to this PR.

- [x] [Unit or functional tests](https://www.elastic.co/guide/en/kibana/master/development-tests.html) were updated or added to match the most common scenarios


## Release note

Adds new `bulkEdit` method to alerting rulesClient and internal _bulk_edit API, that allow bulk editing of rules.

vitaliidm commented:
@elasticmachine merge upstream

XavierM commented on src/core/server/saved_objects/service/lib/aggregations/validation.ts:
```diff
@@ -181,11 +181,18 @@ const recursiveRewrite = (
-      : value;
+
+    let newValue = value;
+    if (rewriteValue) {
```
nit -> to be more readable, will that work
```ts
if (rewriteValue) {
      newValue = validateAndRewriteAttributePath(value, nestedContext);
} else  if (isArray(value)) {
     newValue = value.map((v) => recursiveRewrite(v, nestedContext, parents));
} else if (isPlainObject(value)) {
     newValue = recursiveRewrite(value, nestedContext, [...parents, key]);
}
```

vitaliidm replied:
done

XavierM commented on x-pack/plugins/alerting/server/rules_client/rules_client.ts:
```diff
@@ -179,6 +208,56 @@ export interface FindOptions extends IndexType {
+
+export interface BulkEditOptions<Params extends RuleTypeParams> {
+  filter: string | KueryNode;
+  actions: BulkEditActionRule[];
```
nit -> it might be a good idea to call it `editActions` to avoid confusing with all of our actions terminology

vitaliidm replied:
done

XavierM commented on x-pack/plugins/alerting/server/rules_client/rules_client.ts:
```diff
@@ -1225,6 +1307,231 @@ export class RulesClient {
+              ),
+            };
+            for (const action of actions) {
+              if (action.field === 'actions') {
```
nit -> I think it will be more readable to convert it to switch

vitaliidm replied:
done

XavierM commented on x-pack/plugins/alerting/server/rules_client/rules_client.ts:
```diff
@@ -1225,6 +1307,231 @@ export class RulesClient {
+                ruleActions = this.applyBulkEditAction(action, ruleActions);
+              }
+
+              attributes = this.applyBulkEditAction(action, attributes);
```
Forget `else`
```suggestion
     else {
              attributes = this.applyBulkEditAction(action, attributes);
    }
```

vitaliidm replied:
converted to switch https://github.com/elastic/kibana/pull/126904#discussion_r834609489

XavierM commented on x-pack/plugins/alerting/server/rules_client/rules_client.ts:
```diff
@@ -1225,6 +1307,231 @@ export class RulesClient {
+      await this.authorization.ensureAuthorized({
+        ruleTypeId: ruleType,
+        consumer,
+        operation: WriteOperations.Update,
```
we need to create a new `WriteOperations.BulkEdit`, you can use this PR to see what you need to update https://github.com/elastic/kibana/pull/128163

XavierM reviewed (commented):
Looking good and can you please add api integration test here `x-pack/test/alerting_api_integration/security_and_spaces/tests/alerting` and `x-pack/test/alerting_api_integration/spaces_only/tests/alerting`

mikecote reviewed (commented):
Looking forward to the bulk update API, looking good! I left a few pieces of feedback, let me know if you have any questions.

vitaliidm commented on x-pack/plugins/alerting/server/rules_client/rules_client.ts:
```diff
@@ -1100,6 +1176,224 @@ export class RulesClient {
+      throw Error('Aggregation by alertTypeId|consumer is empty');
+    }
+
+    await pMap(buckets, async ({ key: [ruleType, consumer] }) => {
```
initially, I didn't put it there, as it seemed there are shouldn't be a lot of concurrent actions based on aggregation results. But it might be good thing to put it there.

vitaliidm commented on x-pack/plugins/alerting/server/rules_client/rules_client.ts:
```diff
@@ -1100,6 +1176,224 @@ export class RulesClient {
+            );
+
+            // create API key
+            const username = await this.getUserName();
```
Good catch, thanks. Moved above of `for...of` loop

vitaliidm commented on x-pack/plugins/alerting/server/rules_client/rules_client.ts:
```diff
@@ -1100,6 +1176,224 @@ export class RulesClient {
+            const apiKeyAttributes = this.apiKeyAsAlertAttributes(createdAPIKey, username);
+
+            // get notifyWhen
+            const notifyWhen = getAlertNotifyWhenType(
```
I can see that attributes.notifyWhen can be `null` and this function effectively eliminates null from it.
notifyWhen comes from rules attributes received from results of createPointInTimeFinderAsInternalUser
```typescript
    for await (const response of rulesFinder.find()) {
      await pMap(
        response.saved_objects,
        async (rule) => {
``

How we would be able to make sure it's not null and not calling `getAlertNotifyWhenType`?

mikecote replied:
Ah, this might be a problem with our TypeScript definitions. The existing saved-objects should all have `notifyWhen` set. In 7.11 we put a migration to set a value to all existing documents [ref](https://github.com/elastic/kibana/blob/45c4e7dac1390df210d211fab57a27415be7d93e/x-pack/plugins/alerting/server/saved_objects/migrations.ts#L95) and then set `notifyWhen` via the create API for backwards compatibility when it's not provided.

vitaliidm replied:
So, we can remove `null` here https://github.com/elastic/kibana/blob/main/x-pack/plugins/alerting/server/types.ts#L247

Or leave this `getAlertNotifyWhenType`  as it is?

mikecote replied:
After taking a look, let's leave it as is 👍

mikecote commented on x-pack/plugins/alerting/server/rules_client/rules_client.ts:
```diff
@@ -161,6 +179,56 @@ export interface FindOptions extends IndexType {
+  'actions' | 'schedule' | 'tags' | 'throttle' | 'notifyWhen'
+>;
+
+export type BulkEditActionRule =
```
I like the proposal, I'm ok to go with `operation` to remove the potential confusion 👍

vitaliidm replied:
renamed to `operation`

elasticmachine commented:
Pinging @elastic/response-ops (Team:ResponseOps)

azasypkin commented on x-pack/plugins/encrypted_saved_objects/README.md:
```diff
@@ -99,6 +99,21 @@ const savedObjectWithDecryptedContent =  await esoClient.getDecryptedAsInternalU
 required if Saved Object was created within a non-default space.
 
+Alternative option is using `createPointInTimeFinderAsInternalUser` API method, that can be used to help page through large sets of saved objects.
+Its interface matches interface of corresponding Saved Objects API `createPointInTimeFinder` method :
```
__nit:__
```suggestion
Its interface matches interface of the corresponding Saved Objects API `createPointInTimeFinder` method:
```

azasypkin commented on x-pack/plugins/encrypted_saved_objects/README.md:
````diff
@@ -99,6 +99,21 @@ const savedObjectWithDecryptedContent =  await esoClient.getDecryptedAsInternalU
+Its interface matches interface of corresponding Saved Objects API `createPointInTimeFinder` method :
+
+```typescript
+      const finder = await this.encryptedSavedObjectsClient.createPointInTimeFinderAsInternalUser({
+        filter,
+        type: 'my-saved-object-type',
+        perPage: 1000,
+      });
+
+      for await (const response of finder.find()) {
+        // process response
+      }
````
__nit:__ (to be consistent with the rest of the examples)
```suggestion
const finder = await this.encryptedSavedObjectsClient.createPointInTimeFinderAsInternalUser({
  filter,
  type: 'my-saved-object-type',
  perPage: 1000,
});

for await (const response of finder.find()) {
  // process response
}
```

azasypkin commented on x-pack/plugins/encrypted_saved_objects/server/mocks.ts:
```diff
@@ -25,18 +25,20 @@ function createEncryptedSavedObjectsSetupMock(
   return {
     getDecryptedAsInternalUser: jest.fn(),
+    createPointInTimeFinderAsInternalUser: jest.fn(),
+    bulkGetDecryptedAsInternalUser: jest.fn(),
```
__nit:__ looks like a leftover
```suggestion
```

azasypkin commented on x-pack/plugins/encrypted_saved_objects/server/mocks.ts:
```diff
@@ -25,18 +25,20 @@ function createEncryptedSavedObjectsSetupMock(
   return {
     isEncryptionError: jest.fn(),
-    getClient: jest.fn((opts) => createEncryptedSavedObjectsClienttMock(opts)),
+    getClient: jest.fn((opts) => createEncryptedSavedObjectsClientMock(opts)),
```
:pray:

azasypkin reviewed (commented):
Sorry, didn't manage to complete review today and will finish it tomorrow, just wanted to dump a few minor comments I have so far.

![image](https://user-images.githubusercontent.com/1713708/163220999-c593f922-04e1-44c5-9d11-a09611d1402a.png)

Would you mind adding JSDocs for these 3 new public APIs exposed from ESO plugin?

pmuellr commented:
Just took a quick peek at this PR, not a review of any kind.  I'm wondering if the SO and ESO changes should be split into separate PRs - I'd guess they are needed for the bulk edit functionality, but seems both sets of changes also look useful outside of alerting, so it maybe it would be a little cleaner to separate them.  Not a blocker of any kind though ...

azasypkin commented on x-pack/plugins/encrypted_saved_objects/server/saved_objects/index.ts:
```diff
@@ -96,6 +106,38 @@ export function setupSavedObjects({
+
+      createPointInTimeFinderAsInternalUser: async <T = unknown, A = unknown>(
+        findOptions: SavedObjectsCreatePointInTimeFinderOptions,
+        dependencies?: SavedObjectsCreatePointInTimeFinderDependencies
```
__question:__ I tried to find any info on the `dependencies` parameter to understand the implications of supporting it in ESO, but couldn't. @elastic/kibana-core what `SavedObjectsCreatePointInTimeFinderDependencies` in `createPointInTimeFinder` API is for? I'm wondering if it provides any reasonable benefits to ESO consumers or we can ignore it if it's more an internal Core thing.

vitaliidm replied:
I can see, here is only one property available

```typescript
export interface SavedObjectsCreatePointInTimeFinderDependencies {
  client: Pick<SavedObjectsClientContract, 'find' | 'openPointInTimeForType' | 'closePointInTime'>;
}
```
If I'm not mistaken, looking at code, SO client used here:

```
        const [internalRepository, typeRegistry] = await internalRepositoryAndTypeRegistryPromise;
        const finder = internalRepository.createPointInTimeFinder<T, A>(findOptions, dependencies);
```

is wrapped by EncryptedSavedObjectsClientWrapper? 

which at the same time implemented as:

```typescript
  public createPointInTimeFinder<T = unknown, A = unknown>(
    findOptions: SavedObjectsCreatePointInTimeFinderOptions,
    dependencies?: SavedObjectsCreatePointInTimeFinderDependencies
  ) {
    return this.options.baseClient.createPointInTimeFinder<T, A>(findOptions, {
      client: this,
      // Include dependencies last so that subsequent SO client wrappers have their settings applied.
      ...dependencies,
    });
  }

```

Looks like, this is an ability to add some wrapper on top of SO client.

azasypkin commented on x-pack/plugins/encrypted_saved_objects/server/saved_objects/index.ts:
```diff
@@ -96,6 +106,38 @@ export function setupSavedObjects({
+
+        async function* encryptedFinder() {
+          for await (const res of finderAsyncGenerator) {
+            const encryptedSavedObjects = await pMap(res.saved_objects, async (savedObject) => ({
```
__suggestion:__ Inside of the `pMap` callback we can check `service.isRegistered(savedObject.type)` and return `savedObject` as is if it's not registered with ESO eliminating a slight overhead of a `service.decryptAttributes`.

Alternatively we could filter out not registered types directly from the `findOptions.type` filter before making the request, but that would make this API more convoluted and inconsistent with the existing `getDecryptedAsInternalUser`.

azasypkin replied:
> something like this? 

Yes! Just in case filter includes non-encrypted SO types too.

> Would it make sense to put similar logic for that method as well?

Yeah, it makes sense, and it'd be great if you can update `getDecryptedAsInternalUser` with this logic as well, thanks. I just didn't want to expand the scope of the PR unless you're up for it.

> I think, that wouldn't work for my case, as I need all results from this search query. I this case, I understand, SO with unregistered types would be omitted. I guess, sticking with you previous suggestion would be the best option

I assumed that consumers would likely call this API only with encrypted SO types, but I agree it might not always be the case. Let's stick with the first option :+1:

vitaliidm replied:
implemented

azasypkin commented on x-pack/plugins/encrypted_saved_objects/server/saved_objects/index.ts:
```diff
@@ -96,6 +106,38 @@ export function setupSavedObjects({
+
+        async function* encryptedFinder() {
+          for await (const res of finderAsyncGenerator) {
+            const encryptedSavedObjects = await pMap(res.saved_objects, async (savedObject) => ({
```
__potential issues:__ 

1. How does `pMap` work when we have thousands of saved objects? Does it try to process everything in parallel? If so, I'm afraid it can quickly make Kibana unusable due to high cost of concurrent decryption operations. Did you have a chance to stress-test this a bit? If there a way to control how much tasks it can handle in parallel, the tests can give us a ballpark number.

2. What is the consumer (alerting) expectation if some of the saved objects cannot be decrypted (e.g. wrong encryption key)? How does API behave in this case? I think it makes sense to have a bunch of integration tests for this new API similar (but not limited to) to what we have for [ getDecryptedAsInternalUser](https://github.com/elastic/kibana/blob/main/x-pack/test/encrypted_saved_objects_api_integration/tests/encrypted_saved_objects_api.ts#L363).

vitaliidm replied:
> How does pMap work when we have thousands of saved objects? Does it try to process everything in parallel?
Yes, it would process everything concurrently. But there is a possibility to limit it by setting `concurrency` in pMap call. I put it to 50(usual number we use async operations).

I've measured time it takes: ~ 158.960868999362ms for 100 SO with attributes that get decrypted when Kibana is relatively unloaded, and can take more time under heavy load(~ 500ms)

It takes 100 SO in batch, because I request 100 SO per page through this API

```typescript
    const rulesFinder =
      await this.encryptedSavedObjectsClient.createPointInTimeFinderAsInternalUser<RawRule>({
        filter,
        type: 'alert',
        perPage: 100,
        ...(this.namespace ? { namespaces: [this.namespace] } : undefined),
      });
```

If I understand correctly this service.decryptAttributes is non-blocking? Is that correct?

>What is the consumer (alerting) expectation if some of the saved objects cannot be decrypted (e.g. wrong encryption key)? How does API behave in this case? I think it makes sense to have a bunch of integration tests for this new API similar (but not limited to) to what we have for [getDecryptedAsInternalUser](https://github.com/elastic/kibana/blob/main/x-pack/test/encrypted_saved_objects_api_integration/tests/encrypted_saved_objects_api.ts#L363).

Expectation is similar to failure in case of using  `getDecryptedAsInternalUser`: if expected decoded property(apiKey in our case) is empty, we will just not put this apiKey for further invalidation.
Thanks for pointing out integration tests for existing method. I will look to add some.

azasypkin replied:
> Yes, it would process everything concurrently. But there is a possibility to limit it by setting concurrency in pMap call. I put it to 50(usual number we use async operations).

Got it, thanks. 50 sounds like a reasonable number to me.

> I've measured time it takes: ~ 158.960868999362ms for 100 SO with attributes that get decrypted when Kibana is relatively unloaded, and can take more time under heavy load(~ 500ms)

Thanks for measuring. I was mostly worried about how much stress it can put on Kibana with many concurrent tasks since decryption can be quite CPU intensive. Limiting concurrency "level" to 50 should alleviate this concern though.

> If I understand correctly this service.decryptAttributes is non-blocking? Is that correct?

Yes, that's correct.

> Expectation is similar to failure in case of using getDecryptedAsInternalUser: if expected decoded property(apiKey in our case) is empty, we will just not put this apiKey for further invalidation.

If I remember correctly `getDecryptedAsInternalUser` will throw if it cannot decrypt any of the attributes, so that consumer can know that decryption for the specific SO failed.

But in case of `createPointInTimeFinderAsInternalUser`, what would happen if decryption fails only for _some_ SOs? Will `createPointInTimeFinderAsInternalUser` throw or will "failed" SOs just be excluded from the resulting list of SOs? If excluded, then how would consumer know about these failed SOs?

vitaliidm replied:
> I think it makes sense to have a bunch of integration tests for this new API similar (but not limited to) to what we have for [getDecryptedAsInternalUser](https://github.com/elastic/kibana/blob/main/x-pack/test/encrypted_saved_objects_api_integration/tests/encrypted_saved_objects_api.ts#L363).

implemented

> But in case of createPointInTimeFinderAsInternalUser, what would happen if decryption fails only for some SOs? Will createPointInTimeFinderAsInternalUser throw or will "failed" SOs just be excluded from the resulting list of SOs? If excluded, then how would consumer know about these failed SOs?

So, I've changed here logic a bit, thanks for pointing out that in some cases function decryptAttributes can throw.
To mitigate this and still return SO in list of saved objects and error , I wrapped it in try/catch, if it fails I enrich SO with error and also strip attributes.
```typescript
               try {
                  return {
                    ...savedObject,
                    attributes: (await service.decryptAttributes(
                      descriptor,
                      savedObject.attributes as Record<string, unknown>
                    )) as T,
                  };
                } catch (error) {
                  // catch error and enrich SO with it, return stripped attributes. Then consumer of API can decide either proceed
                  // with only unsecured properties or stop when error happens
                  const { attributes: strippedAttrs } = await service.stripOrDecryptAttributes(
                    descriptor,
                    savedObject.attributes as Record<string, unknown>
                  );
                  return {
                    ...savedObject,
                    attributes: strippedAttrs as T,
                    error: { ...error, message: `Decryption error: "${error.message}"` },
                  };
                }

```

Initially, I thought maybe I have to use `stripOrDecryptAttributes` instead of `decryptAttributes`, but I've noticed in tests: `stripOrDecryptAttributes` doesn't return all properties, hence this implementation.
For reference, loggings from functional test case `'#createPointInTimeFinderDecryptedAsInternalUser decrypts and returns all attributes'`

```bash
 decryptAttributes {
   publicProperty: 'R@)Jo#y6',
   publicPropertyStoredEncrypted: 'ahx$qhD]*RM]^f',
   privateProperty: '2D5tF)Y8Ol4#XG',
   publicPropertyExcludedFromAAD: '[x$4m82T'
 }


 stripOrDecryptAttributes {
   attributes: {
     publicProperty: 'R@)Jo#y6',
     publicPropertyExcludedFromAAD: '[x$4m82T',
     publicPropertyStoredEncrypted: 'ahx$qhD]*RM]^f'
   }
 }
```

azasypkin reviewed (commented):
Left a few more questions/notes. The implementation looks good to me overall, but I want us to be a bit more careful about the potential edge cases and add a few API integration tests as well.

vitaliidm commented:
> Just took a quick peek at this PR, not a review of any kind. I'm wondering if the SO and ESO changes should be split into separate PRs - I'd guess they are needed for the bulk edit functionality, but seems both sets of changes also look useful outside of alerting, so it maybe it would be a little cleaner to separate them. Not a blocker of any kind though ...

Thanks @pmuellr, I think it makes sense. Initially, when started working on it, wasn't clear for us it would require extension to core and ESO. As it already reviewed by those teams, probably make sense to leave as it is, and for the future: split similar large PRs by plugins/team responsibility. What do you think?

jportner commented:
@vitaliidm: @azasypkin is going on vacation, so when you have implemented the suggested changes, please request a review from me directly 🙌

pmuellr commented:
> As it already reviewed by those teams, probably make sense to leave as it is, and for the future: split similar large PRs by plugins/team responsibility. What do you think?

Ya, that's fine.

vitaliidm commented:
@elasticmachine merge upstream

vitaliidm commented:
> @vitaliidm: @azasypkin is going on vacation, so when you have implemented the suggested changes, please request a review from me directly 🙌

Hey @jportner , I've addressed all feedback from Aleh, please take a look when you a free minute

kibanamachine commented:
expected head sha didn’t match current head ref.

azasypkin commented:
__ACK:__ reviewing...

mikecote commented on x-pack/plugins/alerting/server/routes/bulk_edit_rules.ts:
```diff
@@ -0,0 +1,87 @@
+        body: bodySchema,
+      },
+    },
+    router.handleLegacyErrors(
```
We should wrap this with `handleDisabledApiKeysError`?

vitaliidm replied:
Yes, wrapped it. Thanks

mikecote commented on x-pack/plugins/alerting/server/routes/bulk_edit_rules.ts:
```diff
@@ -0,0 +1,87 @@
+        const rulesClient = (await context.alerting).getRulesClient();
+        const { filter, operations, ids } = req.body;
+
+        const bulkEditResults = await rulesClient.bulkEdit({
```
**question:** should we wrap this in a try/catch to handle `RuleTypeDisabledError`. Does the bulkEdit throw `RuleTypeDisabledError` if ever the user attempts to update a rule that requires a higher license level than they currently have?

vitaliidm replied:
You are right, this part is missing. Added it. Thanks

ymao1 commented on x-pack/plugins/alerting/server/rules_client/rules_client.ts:
```diff
@@ -1258,6 +1350,284 @@ export class RulesClient {
+              ruleParams,
+              ruleType.validate?.params
+            );
+            const validatedMutatedAlertTypeParams = validateMutatedRuleTypeParams(
```
Why do we have a separate validation function for updated rule type params? Does this leave it open for params that are invalid but valid according to this separate "mutated param validation" to get saved?

vitaliidm replied:
Here is a link to technical design decision made about it https://github.com/elastic/kibana/issues/124715#issuecomment-1066092827 (second paragraph)

> First of all, we think it will be good idea to add a new property in [AlertTypeParamsValidator](https://github.com/elastic/kibana/blob/main/x-pack/plugins/alerting/server/types.ts#L124-L126) to allow rule type to manage their own validation around their params. We think validate is more here to validate the schema of the params object but validateMutatedParams will be there to do more logical validation for the need of the rule type creator.

ymao1 commented on x-pack/plugins/alerting/tsconfig.json:
```diff
@@ -12,7 +12,7 @@
     "public/**/*",
     "common/*"
-  ],
+, "server/rules_client/lib/.ts"  ],
```
nit: formatting for this is showing up kind of funky :)

vitaliidm replied:
Tbh, I'm not sure how it slipped here. I removed it, there was no intention for me to add it

ymao1 commented on x-pack/test/alerting_api_integration/spaces_only/tests/alerting/bulk_edit.ts:
```diff
@@ -0,0 +1,147 @@
+import { FtrProviderContext } from '../../../common/ftr_provider_context';
+
+// eslint-disable-next-line import/no-default-export
+export default function createUpdateTests({ getService }: FtrProviderContext) {
```
Can we add a functional test for adding an action to ensure that all the reference extracting/injecting works as expected?

vitaliidm replied:
I think I already have tests that check adding actions here: 
https://github.com/elastic/kibana/blob/2e6d918e3c7694c2e129c9f272ea3af8ec75b77e/x-pack/test/alerting_api_integration/security_and_spaces/group1/tests/alerting/bulk_edit.ts#L112-L119

It was done in `security_and_spaces`, not in `spaces_only` to follow the same fashion as similar tests for `update` methos

mikecote commented on x-pack/plugins/alerting/server/rules_client/rules_client.ts:
```diff
@@ -180,6 +211,63 @@ export interface FindOptions extends IndexType {
   filter?: string;
 }
 
+type BulkEditFields = keyof Pick<Rule, 'actions' | 'schedule' | 'tags' | 'throttle' | 'notifyWhen'>;
```
Would this be correct? given the comment below.

```suggestion
type BulkEditFields = keyof Pick<Rule, 'actions' | 'tags'>;
```

vitaliidm replied:
Yeah, would be. I've added this change.

mikecote replied:
Looks good, thanks!

mikecote commented on x-pack/plugins/alerting/server/rules_client/rules_client.ts:
```diff
@@ -1258,6 +1350,284 @@ export class RulesClient {
+
+            const ruleType = this.ruleTypeRegistry.get(rule.attributes.alertTypeId);
+
+            let attributes = cloneDeep(rule.attributes);
```
**question:** is there a place mutating `attributes`? Using `cloneDeep` can be CPU intensive and block the event loop for some time when done at scale.

vitaliidm replied:
attributes get mutated in `applyBulkEditOperation` when we apply operation to it.

```typescript
            for (const operation of operations) {
              switch (operation.field) {
                case 'actions':
                  await this.validateActions(ruleType, operation.value);
                  ruleActions = applyBulkEditOperation(operation, ruleActions);
                  break;
                default:
                  attributes = applyBulkEditOperation(operation, attributes);
              }
            }
```

At this point the only original part of rule needed here(and it's params):

```typescript
            const validatedMutatedAlertTypeParams = validateMutatedRuleTypeParams(
              validatedAlertTypeParams,
              rule.attributes.params,
              ruleType.validate?.params
            );
```

I think in theory, we can do only copy of  `rule.attributes.params` then. 

What can be implications if we mutate rules.attributes instead of copying them for use in `applyBulkEditOperation` in future? Do you think we should be safe to allow mutating them?

mikecote commented on x-pack/plugins/alerting/server/rules_client/rules_client.ts:
```diff
@@ -2353,6 +2723,51 @@ export class RulesClient {
     return alertAttributes;
   }
+
+  private applyBulkEditOperation<R extends object>(operation: BulkEditOperation, rule: R) {
```
**nit:** could this be moved to a separate file?

vitaliidm replied:
moved to a separate file. Thanks for suggestion

mikecote reviewed (commented):
Looking good! I finished my first pass to get this PR reviewed. I'll review tests, pull down, etc next time but wanted to get my questions / some feedback out first.

vitaliidm commented on x-pack/plugins/alerting/server/rules_client/rules_client.ts:
```diff
@@ -1258,6 +1350,284 @@ export class RulesClient {
+      async ({ key: [ruleType, consumer] }) => {
+        this.ruleTypeRegistry.ensureRuleTypeEnabled(ruleType);
+
+        await this.authorization.ensureAuthorized({
```
Good catch! Thank you, I've added audit of the error

vitaliidm commented on x-pack/plugins/alerting/server/rules_client/rules_client.ts:
```diff
@@ -1258,6 +1350,284 @@ export class RulesClient {
+            // validate schedule interval
+            if (attributes.schedule.interval) {
+              const intervalInMs = parseDuration(attributes.schedule.interval as string);
+              if (intervalInMs < this.minimumScheduleIntervalInMs) {
```
thanks for letting know about recent changes in interval validation. I've updated it

vitaliidm commented on x-pack/plugins/alerting/tsconfig.json:
```diff
@@ -12,7 +12,7 @@
     "public/**/*",
     "common/*"
-  ],
+, "server/rules_client/lib/.ts"  ],
```
https://github.com/elastic/kibana/pull/126904#discussion_r869049990

vitaliidm commented on x-pack/plugins/alerting/server/rules_client/rules_client.ts:
```diff
@@ -1258,6 +1350,284 @@ export class RulesClient {
+              notifyWhen,
+              updatedBy: username,
+              updatedAt: new Date().toISOString(),
+            });
```
This part was done before mapped_params were introduced, so I guess would be a fair to add it here. Although we don't bulk edit any of mapped params so far.

Added this piece and test for it. Thanks for suggesting it

vitaliidm commented on x-pack/plugins/alerting/server/invalidate_pending_api_keys/bulk_mark_api_keys_for_invalidation.ts:
```diff
@@ -0,0 +1,35 @@
+/*
```
Done. Have removed `markApiKeyForInvalidation` and its use from other methods

vitaliidm commented:
@elasticmachine merge upstream

mikecote commented on x-pack/plugins/alerting/server/rules_client/lib/apply_bulk_edit_operation.ts:
```diff
@@ -0,0 +1,53 @@
+    return arr.filter((item) => !itemsSet.has(item));
+  };
+
+  switch (operation.operation) {
+    case 'set':
+      set(rule, operation.field, operation.value);
+      break;
+
+    case 'add':
+      set(
+        rule,
+        operation.field,
+        addItemsToArray<AddItemToArray>(get(rule, operation.field) ?? [], operation.value)
+      );
+      break;
+
+    case 'delete':
+      set(
+        rule,
+        operation.field,
+        deleteItemsFromArray(get(rule, operation.field) ?? [], operation.value)
+      );
+      break;
+  }
```
**question:** usage of `set`, will this make us vulnerable to prototype pollution?

mikecote reviewed (approved):
Tested locally, changes LGTM!

vitaliidm commented:
> Tested locally, changes LGTM!

Thanks, @mikecote 
One question from me: I've revisited a bit logic for API keys invalidation

1. To invalidate newly generated API key even if bulkUpdate operation fails
2. Not to invalidate old key, if rule wasn't updated due to conflict for example

Here is a commit. Please let me know if it makes sense.
https://github.com/elastic/kibana/pull/126904/commits/0a34da49e49036fd20104ea8e998383583987158

mikecote commented:
> > Tested locally, changes LGTM!
> 
> Thanks, @mikecote 
> One question from me: I've revisited a bit logic for API keys invalidation
> 
> 1. To invalidate newly generated API key even if bulkUpdate operation fails
> 2. Not to invalidate old key, if rule wasn't updated due to conflict for example
> 
> Here is a commit. Please let me know if it makes sense.
> https://github.com/elastic/kibana/pull/126904/commits/0a34da49e49036fd20104ea8e998383583987158
> 
> 

Yup the change makes sense and looks good 👍 , thanks!
